cli/eth.go

```diff
@@ -0,0 +1,178 @@
+		EthGetAddressCmd,
+		EthCallSimulateCmd,
+	},
+}
+
+func ethAddrFromFilecoinAddress(ctx context.Context, addr address.Address, fnapi v0api.FullNode) (ethtypes.EthAddress, address.Address, error) {
+	var faddr address.Address
+	var err error
+
+	switch addr.Protocol() {
+	case address.BLS, address.SECP256K1:
+		faddr, err = fnapi.StateLookupID(ctx, addr, types.EmptyTSK)
+		if err != nil {
+			return ethtypes.EthAddress{}, addr, err
+		}
+	case address.Actor, address.ID:
+		faddr, err = fnapi.StateLookupID(ctx, addr, types.EmptyTSK)
+		if err != nil {
+			return ethtypes.EthAddress{}, addr, err
+		}
+		fAct, err := fnapi.StateGetActor(ctx, faddr, types.EmptyTSK)
+		if err != nil {
+			return ethtypes.EthAddress{}, addr, err
+		}
+		if fAct.Address != nil && (*fAct.Address).Protocol() == address.Delegated {
+			faddr = *fAct.Address
+		}
+	case address.Delegated:
+		faddr = addr
+	default:
+		return ethtypes.EthAddress{}, addr, xerrors.Errorf("Filecoin address doesn't match known protocols")
+	}
+
+	ethAddr, err := ethtypes.EthAddressFromFilecoinAddress(faddr)
+	if err != nil {
+		return ethtypes.EthAddress{}, addr, err
+	}
+
+	return ethAddr, faddr, nil
+}
+
+var EthGetAddressCmd = &cli.Command{
+	Name:      "stat",
+	Usage:     "Print eth/filecoin addrs and code cid",
+	ArgsUsage: "",
+	Flags: []cli.Flag{
+		&cli.StringFlag{
+			Name:  "filaddr",
+			Value: "",
+			Usage: "Filecoin address",
+		},
+		&cli.StringFlag{
+			Name:  "ethaddr",
+			Value: "",
+			Usage: "Ethereum address",
+		},
+	},
+	Action: func(cctx *cli.Context) error {
+
+		filaddr := cctx.String("filaddr")
+		ethaddr := cctx.String("ethaddr")
+
+		var faddr address.Address
+		var eaddr ethtypes.EthAddress
+
+		api, closer, err := GetFullNodeAPI(cctx)
+		if err != nil {
+			return err
+		}
+		defer closer()
+		ctx := ReqContext(cctx)
+
+		if filaddr != "" {
+			addr, err := address.NewFromString(filaddr)
+			if err != nil {
+				return err
+			}
+			eaddr, faddr, err = ethAddrFromFilecoinAddress(ctx, addr, api)
+			if err != nil {
+				return err
+			}
+		} else if ethaddr != "" {
+			addr, err := hex.DecodeString(ethaddr)
+			if err != nil {
+				return err
+			}
+			copy(eaddr[:], addr)
+			faddr, err = eaddr.ToFilecoinAddress()
+			if err != nil {
+				return err
+			}
+		} else {
+			return xerrors.Errorf("Neither filaddr or ethaddr specified")
+		}
+
+		actor, err := api.StateGetActor(ctx, faddr, types.EmptyTSK)
+		if err != nil {
+			return err
+		}
+
+		fmt.Println("Filecoin address: ", faddr)
+		fmt.Println("Eth address: ", eaddr)
+		fmt.Println("Code cid: ", actor.Code.String())
+
+		return nil
+
+	},
+}
+
+var EthCallSimulateCmd = &cli.Command{
+	Name:      "call",
+	Usage:     "Simulate an eth contract call",
+	ArgsUsage: "[from] [to] [params]",
+	Action: func(cctx *cli.Context) error {
+
+		var fromEthAddr ethtypes.EthAddress
+		fromAddr, err := hex.DecodeString(cctx.Args().Get(0))
+		if err != nil {
+			return err
+		}
+		copy(fromEthAddr[:], fromAddr)
```

Comment:
We should use EthAddressFromHex() here.

Reply:
Better actually to use it on line 102. That way we only have 1 function call and don't have to pre-declare `fromEthAddr`

Reply:
sigh. I looked for a helper for a bit but couldn't find it. Thanks for pointing it out